### ISSUE TITLE
Batchnormalization with cpu only

### ISSUE THREAD

**Issue by jbalma** [Microsoft/CNTK#572]
Hi all,

Is it possible to train using BatchNormalization without GPUs? I'm currently hitting this error when trying to run the ResNet examples on a CPU only, at the start of the first minibatch:

---

Starting Epoch 1: learning rate per sample = 0.015625  effective momentum = 0.900000  momentum as time constant = 607.4 samples
BlockRandomizer::StartEpoch: epoch 0: frames [0..966393](first sequence at sample 0), data subset 2 of 32

Starting minibatch loop, DataParallelSGD training (MyRank = 2, NumNodes = 32, NumGradientBits = 32), distributed reading is ENABLED.
~MPIWrapper

[CALL STACK]
[0x2aaab00c3971]    Microsoft::MSR::CNTK::CPUMatrix<float>::  BatchNormalizationForward  (Microsoft::MSR::CNTK::CPUMatrix<float> const&,  Microsoft::MSR::CNTK::CPUMatrix<float> const&,  double,  double,  Microsoft::MSR::CNTK::CPUMatrix<float>&,  Microsoft::MSR::CNTK::CPUMatrix<float>&,  Microsoft::MSR::CNTK::CPUMatrix<float>&,  double,  Microsoft::MSR::CNTK::CPUMatrix<float>&,  Microsoft::MSR::CNTK::CPUMatrix<float>&) const + 0x49
[0x2aaab01507b7]    Microsoft::MSR::CNTK::Matrix<float>::  BatchNormalizationForward  (Microsoft::MSR::CNTK::Matrix<float> const&,  Microsoft::MSR::CNTK::Matrix<float> const&,  double,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&) const + 0x153
[0x2aaab0185b8e]    Microsoft::MSR::CNTK::CntkBatchNormEngine<float>::  ForwardCore  (Microsoft::MSR::CNTK::Matrix<float> const&,  Microsoft::MSR::CNTK::Matrix<float> const&,  Microsoft::MSR::CNTK::Matrix<float> const&,  double,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&) + 0x22
[0x2aaab0185df8]    Microsoft::MSR::CNTK::BatchNormEngine<float>::  Forward  (Microsoft::MSR::CNTK::Matrix<float> const&,  Microsoft::MSR::CNTK::Matrix<float> const&,  Microsoft::MSR::CNTK::Matrix<float> const&,  double,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&,  double,  Microsoft::MSR::CNTK::Matrix<float>&,  Microsoft::MSR::CNTK::Matrix<float>&) + 0x6e
[0x20083d7e]
[0x20145fba]
[0x20147ef6]
[0x201cae0c]
[0x20205dbc]
[0x2020ae36]
[0x2020d9f4]
[0x202140b4]
[0x2008a280]
[0x20027f7e]
[0x200284fc]
[0x200288fd]
[0x2aaad18c1b05]    __libc_start_main                                  + 0xf5
[0x20023909]
## EXCEPTION occurred: Batch normalization training on CPU is not yet implemented.

The Example ResNet configs/ndls here (https://github.com/Microsoft/CNTK/tree/master/Examples/Image/Miscellaneous/ImageNet/ResNet) use a Macros.ndl which defines BatchNormalization without the useCntkEngine flag as described on the wiki (https://github.com/Microsoft/CNTK/wiki/BatchNormalization) and I'm wondering if this is why I'm hitting this error. 

The implementation in Source/Math/Matrix.cpp seems to imply these flags aren't used, but the template for it in ComputationNetworkLib/ComputationNetworkBuilder.cpp has them taken in and defines them with defaulting to useCntkEngine = true and imageLayoutKind = ImageLayoutKind::CHW. 

Can anyone suggest a work around for getting batchnormalization to work during training on CPU? Ideally, I'd like to be able to compare the same set of ResNet networks across CPU-only and GPU systems. I've seen some hints that this might be possible in #161 and also in #97 in the current master, but so far I haven't been able to get passed this.

-Jacob

**Comment by Alexey-Kamenev** [comment]
Hi Jacob,
Batch normalization currently does not support training on CPU, only evaluation. GPU implementation, is, of course, complete.

**Comment by jbalma** [comment]
Thanks Alexey,

I just found where the placeholder is in CPUMatrix for BatchNormalizationBackward(). I should have checked where the condition was getting triggered before asking ;)

It looks very close. Is BatchNormalizationForward() actually completed? I could maybe try to fill in BatchNormalizationBackward() since most of the hard work has already been done in the cuda versions of these routines. I think most of the required parameters are already computed elsewhere so the big addition would be a similar loop to what's in the forward routine with a different gradient. I'm probably overlooking a few caveats/complications though.

Is there a plan to add CPU support for this in the future?

**Comment by Yandychang1** [comment]
I noticed this thread, while looking for an answer for my problem. is there any plans to do batch normalization training on CPU? I am having the same problem...

**Comment by KeDengMS** [comment]
[Fixed](https://github.com/Microsoft/CNTK/commit/1d3d1e8a2770b65a809f6bf7e6cde49047ffc0c6)
